### ISSUE TITLE
Download Java automatically

### ISSUE THREAD

**Issue by lynrayy** [PolyMC/PolyMC#1534]
### Role

Player, new users of launcher

### Suggestion

Launcher should automatically install java for selected minecraft version

### Benefit

Launcher will be more user-friendly with this feature

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Some people are repelled by the fact that you have to manually install and set java in Prism (j1.8 or j17). Therefore, it will be great if the launcher acts as an official launcher, TLauncher, CurseForge and other user-friendly launchers

**Comment by offbeat-stuff** [comment]
thats bad design

**Comment by offbeat-stuff** [comment]
It takes away user freedom, besides installing java is easy

**Comment by offbeat-stuff** [comment]
though the current polymc is not really associated with freedom

**Comment by lynrayy** [comment]
> It takes away user freedom, besides installing java is easy

You can change java path whenever you want

**Comment by offbeat-stuff** [comment]
But If it installed java automatically that would be kindof bad for linux users especially

**Comment by offbeat-stuff** [comment]
It would at least like it to ask for java to install rather than doing it automatically, and not do it if java is already installed and give multiple options for which java to install

**Comment by lynrayy** [comment]
> But If it installed java automatically that would be kindof bad for linux users especially

Who cares about linux users?
Launcher should be USER FRIENDLY, otherwise less people will be use it and recommend to their friends

**Comment by lynrayy** [comment]
> It would at least like it to ask for java to install rather than doing it automatically, and not do it if java is already installed and give multiple options for which java to install

Good idea

**Comment by Ahptt** [comment]
> > But If it installed java automatically that would be kindof bad for linux users especially
> 
> Who cares about linux users? Launcher should be USER FRIENDLY, otherwise less people will be use it and recommend to their friends

I would argue that if you're getting into a custom launcher, you should have the aptitude to install java and read the output log. The argument of forgetting Linux users doesn't make sense either, considering you petition more people to use it. I don't see a point of a pop up or automatic download when the log tells you the error, and from there you're a search away from installing java (for windows users).

**Comment by lynrayy** [comment]
Why most devs so stupid so they can't understand users want easy start playing?

**Comment by CorniiDog** [comment]
> It takes away user freedom, besides installing java is easy

Have you considered the concept of a check mark in the installer to automatically install java if it's not already installed? And you don't need to check the mark if you plan on manually installing java (which is kinda weird but I guess that's what you want for your "freedom")

**Comment by KaspianDev** [comment]
> > But If it installed java automatically that would be kindof bad for linux users especially
> 
> Who cares about linux users? Launcher should be USER FRIENDLY, otherwise less people will be use it and recommend to their friends

Linux users are easily the biggest audience using this launcher, because it has sane packaging unlike multimc.

**Comment by ghost** [comment]
Automatically downloading Java is not good cross-platform practice. It's trivial enough for the average user to install Java, whatever OS they use. We're really not trying to cater to people using a computer for the first time... the average user nowadays is quite competent with using basic OS features.

Instead what we can do is provide a link to Adoptium for MacOS and Windows, and for Linux/BSD provide a page showcasing what packages your distribution needs to install.

**Comment by lynrayy** [comment]
_Automatically downloading Java is not good cross-platform practice_

Fuck you. I want to have good user experience and i don't care about that. Other launchers somehow downloads java automatically. This function is must-have. Just make it optional if you all like OpEnSoUrCe FaN

**Comment by LennyMcLennington** [comment]
> Automatically downloading Java is not good cross-platform practice. It's trivial enough for the average user to install Java, whatever OS they use. We're really not trying to cater to people using a computer for the first time... the average user nowadays is quite competent with using basic OS features.
> 
> Instead what we can do is provide a link to Adoptium for MacOS and Windows, and for Linux/BSD provide a page showcasing what packages your distribution needs to install.

I had a branch to work on java downloading and improved jre management in general but I never completed it. I know it was possible though. You can pretty easily download and use a JRE from the adoptium API. All "cross platform support" involves is passing the correct OS identifier to the adoptium API so you download the correct JRE for your OS. I don't know whether it would be good to do it fully automatically but it's definitely reasonable to have functionality for downloading and managing multiple JREs within the launcher. Your reasoning makes no sense.

**Comment by ghost** [comment]
That's fine on Windows and MacOS. Doing it on Linux installs it separately of the package manager, which isn't the worst thing if it's stored locally in the launcher.

It's fine if we support this as long as we make it clear that it's probably better for you to install it for yourself if you use java for anything else

> Fuck you. I want to have good user experience and i don't care about that. Other launchers somehow downloads java automatically. This function is must-have. Just make it optional if you all like OpEnSoUrCe FaN

My grandma said the same thing to me when I told her to use Gentoo without my help. She's now ascended and is a god among men, the computer queen of the nursing home. Even the IT staff have nothing on her.

**Comment by lynrayy** [comment]
Fuck linux then. Anyway who play games on linux

**Comment by ghost** [comment]
> Fuck linux then. Anyway who play games on linux

Most of the developers of MultiMC, PolyMC, and PrismLauncher

**Comment by lynrayy** [comment]
I am user, not developer. Don't care. I'm bored of stupid e-mail from this topic that contains something like "Oh no, java autoinstall is against freedom".

Closed.

**Comment by ghost** [comment]
Least angry redditor

**Comment by LennyMcLennington** [comment]
> I am user, not developer. Don't care. I'm bored of stupid e-mail from this topic that contains something like "Oh no, java autoinstall is against freedom".
> 
> Closed.

If you don't want notifications then disable notifications, it's called "unsubscribe".

**Comment by LennyMcLennington** [comment]
> That's fine on Windows and MacOS. Doing it on Linux installs it separately of the package manager, which isn't the worst thing if it's stored locally in the launcher.
> 
> It's fine if we support this as long as we make it clear that it's probably better for you to install it for yourself if you use java for anything else

Linux packaging of Java sucks, especially for Minecraft. Most distros don't package all of the java versions that are required to play all versions of Minecraft, especially on modded. Java runtimes are totally portable, so they can just be downloaded to subdirectory of the launcher's data directory. There's no reason to "make it clear that it's probably better....", since the feature would mainly be aimed at people who don't or can't use the package manager to install java (looking at SteamOS, MacOS, Windows), or people who want a java version that isn't packaged by their distro.

**Comment by ghost** [comment]
That's true. The main reason we wouldn't want to support this would be that it uses up too much space, but I realized you can get a terabyte of storage space with 16GB/s read speed for $40. So it doesn't even matter

**Comment by LennyMcLennington** [comment]
> That's true. The main reason we wouldn't want to support this would be that it uses up too much space, but I realized you can get a terabyte of storage space with 16GB/s read speed for $40. So it doesn't even matter

Most minecraft modpacks take up more space than JREs. Anyway, we could put a warning if the user tries to download a JRE whose exact version is already installed system wide, I guess. Even then, it can be good to have a separate installation even if you have the same version installed system wide, in order to be resistant to OS JRE updates breaking compatibility with forge (it's happened before)

**Comment by ghost** [comment]
That would probably be a good idea.

If we really want to go all in then it wouldn't be a terrible idea to have the Java versions stored within the meta servers and mark certain versions as being broken, and have the launcher just install the latest non-broken one without any user prompt in case the latest version is broken.

**Comment by slendidev** [comment]
On the PolyMC onboarding screen, I would add a button that allows you to fetch the latest (or a specific version) of the Adoptium JRE. Those would be user-local installs of Java (for example, on Linux would be `.local/share/polymc/jre-<version>` or Windows in %APPDATA% etc.) You could also have in the settings window a new tab that contains a JRE manager, which would allow you to delete, add or update versions.

**Comment by slendidev** [comment]
Furthermore, if PolyMC does not find any JRE installed, it can open a dialog that asks if it wants to pull the latest JRE from Adoptium.

If Linux users wish to install from the package manager, they can do so and re-open PolyMC with no issue. We do not need to support integration with different package managers as I would imagine that would lead to a lot of issues.

**Comment by ghost** [comment]
If only we had merged https://github.com/PolyMC/PolyMC/pull/156 we wouldn't have to deal with windows. 😔
